core/src/main/java/com/linecorp/armeria/internal/Http2KeepAliveHandler.java

```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    public void onPingAck(long data) throws Http2Exception {
+        final long elapsed = stopwatch.elapsed(TimeUnit.NANOSECONDS);
+        if (!isGoodPingAck(data)) {
+            return;
+        }
+
+        if (shutdownFuture != null) {
+            final boolean isCancelled = shutdownFuture.cancel(false);
+            if (!isCancelled) {
+                logger.debug("{} shutdownFuture cannot be cancelled because of late PING ACK", channel);
+            }
+        }
+        logger.debug("{} Received PING(ACK=1) in {} ns", channel, elapsed);
+        state = State.IDLE;
+    }
+
+    private boolean isGoodPingAck(long data) {
+        if (state != State.PENDING_PING_ACK) {
+            logger.warn("Unexpected PING(ACK=1, DATA={}) received", data);
+            return false;
+        }
+        if (lastPingPayload != data) {
+            logger.warn("Unexpected PING(ACK=1, DATA={}) received, " +
+                    "but expecting PING(ACK=1, DATA={})", data, lastPingPayload);
+            return false;
+        }
+        return true;
+    }
+
+    @VisibleForTesting
+    State state() {
+        return state;
+    }
+
+    @VisibleForTesting
+    long lastPingPayload() {
+        return lastPingPayload;
+    }
+
+    private void closeChannelAndLog() {
+        if (state == State.SHUTDOWN) {
+            return;
+        }
+        logger.debug("{} Closing channel", channel);
+        channel.close().addListener(future -> {
+            if (future.isSuccess()) {
+                logger.debug("{} Closed channel", channel);
+            } else {
+                logger.debug("{} Channel cannot be closed", channel, future.cause());
+            }
+            state = State.SHUTDOWN;
+        });
+    }
+
+    /**
+     * State changes from IDLE -> PING_SCHEDULED -> PENDING_PING_ACK -> IDLE and so on. When the
+     * channel is inactive then the state changes to SHUTDOWN.
+     */
+    enum State {
+        /* Nothing happening, but waiting for IdleStateEvent */
+        IDLE,
+
+        /* PING is scheduled */
+        PING_SCHEDULED,
+
+        /* PING is sent and is pending ACK */
+        PENDING_PING_ACK,
+
+        /* Not active anymore */
+        SHUTDOWN
+    }
+
+    private class PingWriteListener implements ChannelFutureListener {
+
+        @Override
+        public void operationComplete(ChannelFuture future) throws Exception {
+            if (future.isSuccess()) {
+                final EventLoop el = channel.eventLoop();
+                shutdownFuture = el.schedule(shutdownRunnable, pingTimeoutMillis, TimeUnit.MILLISECONDS);
+                state = State.PENDING_PING_ACK;
+                stopwatch.reset().start();
+            } else {
+                // Mostly because the channel is already closed. So ignore and change state to IDLE.
+                // If the channel is closed, we change state to SHUTDOWN on onChannelInactive.
+                logger.debug("{} Channel PING write failed", channel);
```

Comment:
How about not logging if cause is `Exceptions.isExpected()` returns true? Otherwise, we could log with the cause.